packages/markers/src/browser/problem/problem-decorator.ts

```diff
@@ -0,0 +1,117 @@
+        for (const [uri, marker] of new Map(markers.map(m => [new URI(m.uri), m] as [URI, Marker<Diagnostic>])).entries()) {
+            const uriString = uri.toString();
+            result.set(uriString, marker);
+            let parentUri: URI | undefined = uri.parent;
+            while (parentUri && !parentUri.path.isRoot) {
+                const parentUriString = parentUri.toString();
+                const existing = result.get(parentUriString);
+                // Make sure the highest diagnostic severity (smaller number) will be propagated to the container directory.
+                if (existing === undefined || severityCompare(existing, marker) < 0) {
+                    result.set(parentUriString, {
+                        data: marker.data,
+                        uri: parentUriString,
+                        owner: marker.owner,
+                        kind: marker.kind
+                    });
+                    parentUri = parentUri.parent;
+                } else {
+                    parentUri = undefined;
+                }
+            }
+        }
+        return Array.from(result.values());
+
+    }
+
+    protected collectMarkers(): Marker<Diagnostic>[] {
+        return Array.from(this.problemManager.getUris())
+            .map(uri => new URI(uri))
+            .map(uri => this.problemManager.findMarkers({ uri }))
+            .map(markers => markers.sort(severityCompare))
+            .map(markers => markers.shift())
+            .filter(notEmpty);
+    }
+
+    protected toDecorator(marker: Marker<Diagnostic>): DecorationData {
+        const position = IconOverlayPosition.BOTTOM_RIGHT;
+        const icon = this.getOverlayIcon(marker);
+        const color = this.getOverlayIconColor(marker);
+        return {
+            iconOverlay: {
+                position,
+                icon,
+                color
+            }
+        };
+    }
+
+    protected getOverlayIcon(marker: Marker<Diagnostic>): string {
+        const { severity } = marker.data;
+        switch (severity) {
+            case 1: return 'times-circle';
+            case 2: return 'exclamation-circle';
+            case 3: return 'info-circle';
+            default: return 'hand-o-up';
+        }
+    }
+
+    protected getOverlayIconColor(marker: Marker<Diagnostic>): Color {
+        const { severity } = marker.data;
+        switch (severity) {
+            case 1: return 'var(--theia-error-color0)';
+            case 2: return 'var(--theia-warn-color0)';
+            case 3: return 'var(--theia-info-color0)';
+            default: return 'var(--theia-success-color0)';
+        }
+    }
+
+}
+
+// Highest severities (errors) come first, then the others. Undefined severities treated as the last ones.
+const severityCompare = (left: Marker<Diagnostic>, right: Marker<Diagnostic>): number =>
```

Comment:
please put global variables under namespaces, i.e. `ProblemDecorator.severityCompare`

Reply:
Is that required even if it not exported? I can for sure, just asking what the policy is.

Reply:
not, it just looks like if one customize `ProblemDecorator` then this function can be handy

Reply:
Makes sense. I will update it. Thanks!